src/absil/il.fs

```diff
@@ -850,33 +850,60 @@ type ILAttribElem =
   | Array of ILType * ILAttribElem list
 
 type ILAttributeNamedArg =  (string * ILType * bool * ILAttribElem)
-type ILAttribute = 
-    { Method: ILMethodSpec
-      Data: byte[] 
-      Elements: ILAttribElem list}
+
+type IAttribute =
+    abstract Method: ILMethodSpec  
+    abstract Data: byte[] 
+    abstract Elements: ILAttribElem list
+
+    abstract With:
+        ?newMethod: ILMethodSpec *
+        ?newData: byte[] *
+        ?newElements: ILAttribElem list
+            -> IAttribute
+
+
+type IAttributes =
+    abstract AsArray: IAttribute []
+    abstract AsList: IAttribute list
+
+
+[<Sealed>]
+type ILAttribute(method, data, elements) = 
```

Comment:
I've requested that the API use the names `ILAttribute`, partly to completely minimize the code churn and de-emphasize the abstraction being performed here (we do _not_ want to consider arbitrary implementors in this codebase, just implementors that implement contracts very similar to the existing functional-object contracts, e.g. take-no-locks-and-do-no-observable-mutation).  

As a result we need to use a different name here. I think it's probably best to use `ILAttributeImpl` etc.